Refine readme (#749)
ROCmSoftwarePlatform/AMDMIGraphX: README.md

# AMD MIGraphX

AMD MIGraphX is AMD's graph inference engine to accelerate model inference on AMD GPUs. AMD MIGraphX can be used by
installing binaries directly or building from source code.

Note that all the following instructions are based on that ROCm has been installed successfully. ROCm installation
instructions are explained in the [ROCm installation
guide](https://rocmdocs.amd.com/en/latest/Installation_Guide/Installation-Guide.html).

## Installing from binaries

> With ROCm installed correctly, MIGraphX binaries can be installed with the following command:

Reviewer: > MIGraphX binaries can be installed on ubuntu with the following command

Author: mentioned Ubuntu

```
sudo apt update && sudo apt install -y migraphx
```

> then the head files and libs are located at `/opt/rocm/include` and `/opt/rocm/lib`, respectively, which can be

Reviewer: First, it should be header and not head. Secondly, they are not located in /opt/rocm, unless the user adds a symlink. It should say that migraphx is installed under `/opt/roc,-<version>` where <version> is the rocm version.

Author: added.

> included and linked by adding the corresponding folders to the Makefile.

Reviewer: This line should be removed. MIGraphX should be linked using the `migraphx::migraphx_c` cmake target. We should add another sections that says:

> To use MIGraphX's C/C++ API in your cmake project, then set `CMAKE_PREFIX_PATH` to your rocm installation and then do(where `myApp` is the cmake target in your project):

```cmake
find_package(migraphx)
target_link_libraries(myApp migraphx::migraphx_c)
```

Author: Added a section to mention this part.

## Building from source

There are two ways to build the MIGraphX sources. One is installing the dependencies, then using 
cmake to build the source. The other is using the ROCm build tool [rbuild](https://github.com/RadeonOpenCompute/rbuild).
In the following, we will first list the dependencies required to build MIGraphX source code, then describe each of the
two approaches.

> ### List of dependencies

Reviewer: I think this should `Prerequisites`

Author: changed accordingly

The following is a list of dependencies required to build MIGraphX source. This list is also available in the

> requirement files `dev-requirements.txt` and `requirements.txt`.

Reviewer: I wouldn't mention the requirements file. In the future, we will remove rocblas and miopen from the requirements file so it will be incomplete.

* [ROCm cmake modules](https://github.com/RadeonOpenCompute/rocm-cmake) **required**
* [MIOpen](https://github.com/ROCmSoftwarePlatform/MIOpen) for running on the GPU

> * [rocBLAS](https://github.com/ROCmSoftwarePlatform/rocBLAS) for running on the GPU

Reviewer: You are also missing msgpack and json library in the list.

Author: added.

* [HIP](https://github.com/ROCm-Developer-Tools/HIP) for running on the GPU
* [Protobuf](https://github.com/google/protobuf) for reading [onnx](https://github.com/onnx/onnx) files
* [Half](http://half.sourceforge.net/) - IEEE 754-based half-precision floating point library
* [pybind11](https://pybind11.readthedocs.io/en/stable/) - for python bindings

#### Use cmake to build MIGraphX

In this approach, we need to install the dependencies, configure the cmake, and then build the source.

##### Installing the dependencies

> You can manually download and installing the above dependencies one by one. For convience, we provide a shell 

Reviewer: This seems to imply that you can manually install dependencies or use install_prereqs.sh. This should read

> For convenience, the dependencies can be built automatically with rbuild. If you are on ubuntu 18.04 and have sudo access then you can install the dependencies just like in the docker file by calling ./tools/install_prereqs.sh.

Using rbuild should be emphasized and preferred since its portable.

Author: moved rbuild before using cmake..

script [install_prereqs.sh](./tools/install_prereqs.sh) that can automatically install all the above dependencies with
the command 

```./tools/install_prereqs.sh```

(Note: By default, all dependencies are installed at the default location `/usr/local` 
and are accessible by all users. For the default location, `sudo` is required to run the script.
You can also specify a location at which the dependencies are installed with `./tools/install_prereqs.sh $your_loc`.)

##### Building MIGraphX source and install libs

With the above dependencies installed, we can build source as:

1) Go to the project folder and create a build directory:

```
mkdir build
cd build
```

2) Configure the cmake. If the dependencies are installed at the default location `/usr/local`, the command is:

```
CXX=/opt/rocm/llvm/bin/clang++ cmake ..
```
Otherwise, you need to set `-DCMAKE_PREFIX_PATH=$your_loc` to configure the cmake. 

3) Build MIGraphX source code

```
make -j$(nproc)
```

Correctness can be verified as:

```
make -j$(nproc) check
```

MIGraphX libs can be installed as:

```
make install
```

> #### Use the ROCm build tool [rbuild](https://github.com/RadeonOpenCompute/rbuild).

Reviewer: Building with rbuild should be mentioned first before install_prereqs.sh file.

In this approach, we need to install the [rbuild](https://github.com/RadeonOpenCompute/rbuild) first, then use it to
build MIGraphX. rbuild can be installed as (sudo may be needed.):
```
pip3 install https://github.com/RadeonOpenCompute/rbuild/archive/master.tar.gz
```
and pip3 can be installed as `sudo apt update && sudo apt install -y python3-pip`.

> We also need to install [rocm-cmake](https://github.com/RadeonOpenCompute/rocm-cmake) as `sudo apt install -y rocm-cmake`.

Reviewer: More than just `rocm-cmake` is needed. They need to install both `miopen-hip` and `rocblas` as well.

Reviewer: There should be a seperate step that describes which rocm packages need to be installed. Also, this should clarify that these are the packages for ubuntu.

Author: The reason I wrote it in this way is: rocm-cmake is needed before we can run the rbuild. As for the miopen and rocblas, they are installed in the script. I will not mention rocm-cmake here and move it to the script file, so it will be installed in the same way as the miopen and rocblas.

Then MIGraphX can be built as:

```
rbuild build -d depend -B build --cxx=/opt/rocm/llvm/bin/clang++
```

Note that for ROCm3.7 and later release, Ubuntu 18.04 or later releases are needed. Upgrapde to Ubuntu 18.04 can be
done as:

```
sudo apt update

> sudo apt install linux-headers-4.18.0-25-generic linux-image-4.18.0-25-generic  linux-modules-4.18.0-25-generic linux-modules-extra-4.18.0-25-generic -y

Reviewer: I dont think we should add documentation for specific version of rocm. Also, the rocm installation document should cover this.

Author: moved to a wiki page for that.

* [MIOpen](https://github.com/ROCmSoftwarePlatform/MIOpen) for running on the GPU
* [rocBLAS](https://github.com/ROCmSoftwarePlatform/rocBLAS) for running on the GPU
* [HIP](https://github.com/ROCm-Developer-Tools/HIP) for running on the GPU
* [Protobuf](https://github.com/google/protobuf) for reading [onnx](https://github.com/onnx/onnx) files
* [Half](http://half.sourceforge.net/) - IEEE 754-based half-precision floating point library
* [pybind11](https://pybind11.readthedocs.io/en/stable/) - for python bindings
* [JSON](https://github.com/nlohmann/json) - for model serialization to json string format
* [MessagePack](https://msgpack.org/index.html) - for model serialization to binary format

#### Use the ROCm build tool [rbuild](https://github.com/RadeonOpenCompute/rbuild).

In this approach, we need to install the [rbuild](https://github.com/RadeonOpenCompute/rbuild) first, then use it to
build MIGraphX. rbuild can be installed as (sudo may be needed):
```
pip3 install https://github.com/RadeonOpenCompute/rbuild/archive/master.tar.gz
```
and pip3 can be installed as `sudo apt update && sudo apt install -y python3-pip`.

> We also need to install [rocm-cmake](https://github.com/RadeonOpenCompute/rocm-cmake), rocblas, and miopen-hip as 

Reviewer: I think this should be mentioned before installing rbuild. Something:

> To install the rocm dependencies on ubuntu from repo.radeon.com, you can run this `apt` command:
> 
>     sudo apt install rocm-cmake rocblas miopen-hip python3-pip

Then next mention how to install rbuild.

> `sudo apt install -y rocm-cmake rocblas miopen-hip`.

Reviewer: This should just be indented with four spaces and then remove the backticks.

Then MIGraphX can be built as:

```
rbuild build -d depend -B build --cxx=/opt/rocm/llvm/bin/clang++
```

Note that for ROCm3.7 and later releases, Ubuntu 18.04 or later releases are needed. 
Upgrapde to Ubuntu 18.04 is available at [Upgrade Ubuntu to 18.04](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/wiki/Upgrade-to-Ubuntu-18.04-for-ROCM3.7-or-later-releases)

#### Use cmake to build MIGraphX

In this approach, we need to install the prerequisites, configure the cmake, and then build the source.

##### Installing the prerequisites

For convenience, the prerequisites can be built automatically with rbuild as:

> ```rbuild build -d depend --cxx=/opt/rocm/llvm/bin/clang++```, all the prerequisites are in the 

Reviewer: This should be put on a seperate line.

> folder `depend`. 

Reviewer: It also should be mentioned that migraphx will be built in the `build` directory.

If you have sudo access, as an alternative to the rbuild command, you can install the prerequisites just 

> like in the docker file by calling ```./tools/install_prereqs.sh.```

Reviewer: A triple backtick is not needed for code blocks that are inline, just a single backtick will work.

(Note: By default, all dependencies are installed at the default location `/usr/local` 
and are accessible by all users. For the default location, `sudo` is required to run the script.
You can also specify a location at which the dependencies are installed with `./tools/install_prereqs.sh $your_loc`.)

##### Building MIGraphX source and install libs

With the above dependencies installed, we can build source as:

1) Go to the project folder and create a build directory:

```
mkdir build
cd build
```

2) Configure the cmake. If the dependencies are installed at the default location `/usr/local`, the command is:

```
CXX=/opt/rocm/llvm/bin/clang++ cmake ..
```
Otherwise, you need to set `-DCMAKE_PREFIX_PATH=$your_loc` to configure the cmake. 

3) Build MIGraphX source code

```
make -j$(nproc)
```

Correctness can be verified as:

```
make -j$(nproc) check
```

MIGraphX libs can be installed as:

```
make install
```

### Calling MIGraphX APIs
To use MIGraphX's C/C++ API in your cmake project, then set `CMAKE_PREFIX_PATH` to your rocm installation and then do 
```
find_package(migraphx)

> target_link_libraries(myApp migraphx::migraphx_c)

Reviewer 2: In the C++ examples I've done so far, this needed to be migraphx::c, rather than migraphx::migraphx_c, to compile. @pfultz2 can you confirm whether this is the case?

Reviewer: Yes, thats correct!

# AMD MIGraphX

AMD MIGraphX is AMD's graph inference engine to accelerate model inference on AMD GPUs. AMD MIGraphX can be used by
installing binaries directly or building from source code.

In the following, instructions of how to build and install MIGraphX are described with Ubuntu as the OS
(Instructions of installation on other Linux OSes will come later). Note that all the following instructions are based 
on that ROCm has been installed successfully. ROCm installation instructions are explained in the [ROCm installation
guide](https://rocmdocs.amd.com/en/latest/Installation_Guide/Installation-Guide.html).

## Installing from binaries
With ROCm installed correctly, MIGraphX binaries can be installed on Ubuntu with the following command:
```
sudo apt update && sudo apt install -y migraphx
```

> then the header files and libs are installed under `/opt/rocm-<version>`, where `<version>` is the rocm versoin.

Reviewer: Typo: versoin -> version

```

2) Install [rbuild](https://github.com/RadeonOpenCompute/rbuild) (sudo may be required here.)

```
pip3 install https://github.com/RadeonOpenCompute/rbuild/archive/master.tar.gz
```

3) Build MIGraphX source code

```
rbuild build -d depend -B build --cxx=/opt/rocm/llvm/bin/clang++
```

then all the prerequisites are in the folder `depend`, and MIGraphX is built in the `build` directory.

Note that for ROCm3.7 and later releases, Ubuntu 18.04 or later releases are needed. 
Upgrapde to Ubuntu 18.04 is available at [Upgrade Ubuntu to 18.04](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/wiki/Upgrade-to-Ubuntu-18.04-for-ROCM3.7-or-later-releases)

#### Use cmake to build MIGraphX

In this approach, we need to install the prerequisites, configure the cmake, and then build the source.

##### Installing the prerequisites

For convenience, the prerequisites can be built automatically with rbuild as:

```
rbuild build -d depend --cxx=/opt/rocm/llvm/bin/clang++
```

then all the prerequisites are in the folder `depend`, and they can be used in the `cmake` configuration
as `-DCMAKE_PREFIX_PATH=depend`.

> If you have sudo access, as an alternative to the rbuild command, you can install the prerequisites just 

Reviewer: This script also requires ubuntu.

Author: Added a note for this.

included and linked by adding the corresponding folders to the Makefile.

## Building from source

Building MIGraphX from sources needs dependencies, which must be installed before building the source code. In the
following, we first list the dependencies, and then explain two ways of installing them.

### List of dependencies
The following is a list of dependencies required to build MIGraphX from source code. This list is also available in the
requirement files ```dev-requirements.txt``` and ```requirements.txt```.

* [ROCm cmake modules](https://github.com/RadeonOpenCompute/rocm-cmake) **required**
* [MIOpen](https://github.com/ROCmSoftwarePlatform/MIOpen) for running on the GPU
* [rocBLAS](https://github.com/ROCmSoftwarePlatform/rocBLAS) for running on the GPU
* [HIP](https://github.com/ROCm-Developer-Tools/HIP) for running on the GPU
* [Protobuf](https://github.com/google/protobuf) for reading [onnx](https://github.com/onnx/onnx) files
* [Half](http://half.sourceforge.net/) - IEEE 754-based half-precision floating point library
* [pybind11](https://pybind11.readthedocs.io/en/stable/) - for python bindings

> * [ONNX 1.7.0](https://github.com/onnx/onnx) and [Pytest](https://github.com/pytest-dev/pytest) for running the ONN backend

Reviewer: This dependency should not be listed here, as this is only needed to run the tests, so most users will not need this.

Author: OK, I move it to another place.

ONNX 1.7.0 can be installed as ```pip3 install onnx==1.7.0```, and updates to support ONNX version 1.8.0 will come soon. 2)
MIOpen and rocBLAS can be installed as ```sudo apt update && sudo apt install -y miopen-hip rocblas```.

Dependencies can be installed using the ROCm build tool [rbuild](https://github.com/RadeonOpenCompute/rbuild).
### Installing the dependencies
There are two alternative ways to install the above dependencies:

To install rbuild:
```
pip install https://github.com/RadeonOpenCompute/rbuild/archive/master.tar.gz
```
#### Run a shell script [build_prereqs.sh](./tools/build_prereqs.sh).

(Note: 1. You need the sudo to install the above dependencies. 2. All dependencies are installed in default locations in
the system and are accessible by all users.)

#### Use the ROCm build tool [rbuild](https://github.com/RadeonOpenCompute/rbuild).

To build dependencies along with MIGraphX
To install rbuild (sudo may be needed.):
```
rbuild build -d depend --cxx=/opt/rocm/llvm/bin/clang++
pip3 install https://github.com/RadeonOpenCompute/rbuild/archive/master.tar.gz
```

> This builds dependencies in the subdirectory named depend and then builds MIGraphX using these dependencies.

Reviewer 3: This line was confusing. This looks much better, in terms of flow for building and running.

Author: Thanks. I will reword this sentence.

> rbuild develop -d depend -B build --cxx=/opt/rocm/bin/hcc

Reviewer: Hcc is deprecated so I would just remove this from the readme.

Author: If someone use a old GPU like the gfx803, rocm3.3 maybe needed as we mentioned in yesterday's team meeting.

Author: opinions from other guys are appreciated.

Reviewer: > If someone use a old GPU like the gfx803, rocm3.3 maybe needed as we mentioned in yesterday's team meeting.

I think we should just have wiki page to go over using migraphx with older rocm 3.3 then. We only test migraphx with hcc to use ubsan, however once clang supports ubsan then we will drop testing and support for rocm 3.3.

Author: added a wiki page for rocm3.3 at: https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/wiki/Build-MIGraphX-Source-in-ROCm3.3

```

> rbuild develop -d depend -B build --cxx=/opt/rocm/llvm/clang++

Reviewer: This should be `rbuild build` or `rbuild package`. `rbuild develop` is only needed for developers.

Author: I just tried the command "rbuild build", that is the error message I got.
```
Successfully installed msgpack/msgpack-c
-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is Clang 12.0.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /opt/rocm/llvm/bin/clang++
-- Check for working CXX compiler: /opt/rocm/llvm/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:21 (find_package):
  By not providing "FindROCM.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "ROCM", but
  CMake did not find one.

  Could not find a package configuration file provided by "ROCM" with any of
  the following names:

    ROCMConfig.cmake
    rocm-config.cmake

  Add the installation prefix of "ROCM" to CMAKE_PREFIX_PATH or set
  "ROCM_DIR" to a directory containing one of the above files.  If "ROCM"
  provides a separate development package or SDK, be sure it has been
  installed.
```

Author: In my opinion, we should avoid this kind of errors. It is easy for an external user to give up

Reviewer: The user needs to install rocm-cmake from the package manager.

Reviewer: Rocm-cmake should be included in the `apt install` command.

```

This builds dependencies in the folder "depend".

Note that if rocBLAS and MIOpen are installed with the command ```sudo apt install -y rocblas miopen-hip``` as mentioned
above, we can comment out the two lines 
```
ROCmSoftwarePlatform/rocBLAS@abd98a2b48b29326ebaef471630786a548622c06
ROCmSoftwarePlatform/MIOpen@2.4.0
```

> in the file `requirements.txt` (adding a '#' character at the start of each line).

Reviewer: There is no reason to comment out these lines. `rbuild` should automatically ignore these.

Also note that for ROCm3.7 and later release, Ubuntu 18.04 or later releases are needed. Upgrapde to Ubuntu 18.04 can be

> done as:

Reviewer: We dont need to document how to upgrade to ubuntu 18.04, we should just refer to rocm documentation on how to install rocm.

Author: Then we need a link to point to where Ubuntu can be upgraded. Do you have an idea of where we have such information?

```
sudo apt update
sudo apt install linux-headers-4.18.0-25-generic linux-image-4.18.0-25-generic  linux-modules-4.18.0-25-generic linux-modules-extra-4.18.0-25-generic -y
```

### Building MIGraphX source

First create a build directory:

```
mkdir build; 
cd build;
```

Next, configure the cmake. If the script [buiild_prereqs.sh](./tools/build_prereqs.sh) is used to install the
dependencies, then all dependencies are installed at default locations, and MIGraphX can be build as:
* In ROCm3.3:

```
CXX=/opt/rocm/bin/hcc cmake ..
```

* In ROCm3.7 or later releases:

```
CXX=/opt/rocm/llvm/bin/clang++ cmake ..
```

If the above rbuild command was used to build and install the dependencies, then all dependencies are in the folder
"depend", and the `CMAKE_PREFIX_PATH` needs to be set to the same folder "depend" (full path are needed here.), and the

> command is:

Reviewer: `rbuild` already creates a build directory and configures it with cmake(unless you used `rbuild prepare` which the readme doesn't mention).

* ROCM3.3:

```
CXX=/opt/rocm/bin/hcc cmake -DCMAKE_PREFIX_PATH=../depend ..

```
* ROCM3.7 or later releases:

```
CXX=/opt/rocm/llvm/bin/clang++ cmake -DCMAKE_PREFIX_PATH=../depend ..
```

Then we can build MIGraphX source code as:

```
make -j$(nproc)
```

and correctness can be verified as:

```
make -j$(nproc) check
```

MIGraphX libs can be installed as:

```

> sudo make install

Reviewer: Dont put `sudo` here. If the user sets the install prefix to a local path then sudo is nto necessary.

Author: sudo is removed.

# AMD MIGraphX

> AMD MIGraphX is AMD's graph inference engine to accelerate model inference on AMD GPUs. AMD MIGraphX can be used by

Reviewer 4: ```suggestion
AMD MIGraphX is AMD's graph inference engine that accelerates machine learning model inference. AMD MIGraphX can be used by
```

installing binaries directly or building from source code.

In the following, instructions of how to build and install MIGraphX are described with Ubuntu as the OS

> (Instructions of installation on other Linux OSes will come later). Note that all the following instructions are based 

Reviewer 4: ```suggestion
(Instructions of installation on other Linux OSes will come later). Note that all the following instructions assume 
```

> on that ROCm has been installed successfully. ROCm installation instructions are explained in the [ROCm installation

Reviewer 4: ```suggestion
ROCm has been installed successfully. ROCm installation instructions are explained in the [ROCm installation
```

guide](https://rocmdocs.amd.com/en/latest/Installation_Guide/Installation-Guide.html).

## Installing from binaries
With ROCm installed correctly, MIGraphX binaries can be installed on Ubuntu with the following command:
```
sudo apt update && sudo apt install -y migraphx
```

> then the header files and libs are installed under `/opt/rocm-<version>`, where `<version>` is the rocm version.

Reviewer 4: ```suggestion
then the header files and libs are installed under `/opt/rocm-<version>`, where `<version>` is the ROCm version.
```

* [pybind11](https://pybind11.readthedocs.io/en/stable/) - for python bindings
* [JSON](https://github.com/nlohmann/json) - for model serialization to json string format
* [MessagePack](https://msgpack.org/index.html) - for model serialization to binary format

#### Use the ROCm build tool [rbuild](https://github.com/RadeonOpenCompute/rbuild).

In this approach, we use the [rbuild](https://github.com/RadeonOpenCompute/rbuild) build tool to
build MIGraphX. The specific steps are as follows:

1) Install rocm-cmake, pip3, rocblas, and miopen-hip with the command

```
sudo apt update && sudo apt install -y rocm-cmake python3-pip rocblas miopen-hip
```

2) Install [rbuild](https://github.com/RadeonOpenCompute/rbuild) (sudo may be required here.)

```
pip3 install https://github.com/RadeonOpenCompute/rbuild/archive/master.tar.gz
```

3) Build MIGraphX source code

```
rbuild build -d depend -B build --cxx=/opt/rocm/llvm/bin/clang++
```

then all the prerequisites are in the folder `depend`, and MIGraphX is built in the `build` directory.

Note that for ROCm3.7 and later releases, Ubuntu 18.04 or later releases are needed. 

> Upgrapde to Ubuntu 18.04 is available at [Upgrade Ubuntu to 18.04](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/wiki/Upgrade-to-Ubuntu-18.04-for-ROCM3.7-or-later-releases)

Reviewer 4: ```suggestion
Upgrade to Ubuntu 18.04 is available at [Upgrade Ubuntu to 18.04](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/wiki/Upgrade-to-Ubuntu-18.04-for-ROCM3.7-or-later-releases)
```

Also note that you may meet the error of `rbuild: command not found`. It is because rbuild is installed 
at `$HOME/.local/bin`, which is not in `PATH`. You can either export PATH as `export PATH=$HOME/.local/bin:$PATH` 
to add the folder to `PATH` or add the option `--prefix /usr/local` in the pip3 command when installing rbuild.

#### Use cmake to build MIGraphX

> In this approach, we need to install the prerequisites, configure the cmake, and then build the source.

Reviewer 4: ```suggestion
If using this approach, we need to install the prerequisites, configure the cmake, and then build the source.
```

##### Installing the prerequisites

For convenience, the prerequisites can be built automatically with rbuild as:

```
rbuild build -d depend --cxx=/opt/rocm/llvm/bin/clang++
```

then all the prerequisites are in the folder `depend`, and they can be used in the `cmake` configuration
as `-DCMAKE_PREFIX_PATH=depend`.

If you have sudo access, as an alternative to the rbuild command, you can install the prerequisites just 

> like in the docker file by calling `./tools/install_prereqs.sh`.

Reviewer 4: ```suggestion
like in the dockerfile by calling `./tools/install_prereqs.sh`.
```

(Note that this script is for Ubuntu. By default, all prerequisites are installed at the default location `/usr/local` 
and are accessible by all users. For the default location, `sudo` is required to run the script.
You can also specify a location at which the prerequisites are installed with `./tools/install_prereqs.sh $your_loc`.)

##### Building MIGraphX source and install libs

With the above prerequisites installed, we can build source as:

1) Go to the project folder and create a `build` directory:

```
mkdir build
cd build
```

2) Configure the cmake. If the prerequisites are installed at the default location `/usr/local`, the command is:

```
CXX=/opt/rocm/llvm/bin/clang++ cmake ..
```
Otherwise, you need to set `-DCMAKE_PREFIX_PATH=$your_loc` to configure the cmake. 

3) Build MIGraphX source code

```
make -j$(nproc)
```

Correctness can be verified as:

```
make -j$(nproc) check
```

MIGraphX libs can be installed as:

```
make install
```

#### Using docker

> The easiest way to setup the development environment is to use docker. With the docker file, you can build a docker image as:

Reviewer 4: ```suggestion
The easiest way to setup the development environment is to use docker. With the dockerfile, you can build a docker image as:
```

    docker build -t migraphx .

Then to enter the developement environment use `docker run`:

    docker run --device='/dev/kfd' --device='/dev/dri' -v=`pwd`:/code/AMDMIGraphX -w /code/AMDMIGraphX --group-add video -it migraphx

In the docker container, all the required prerequisites are already installed, so users can just go to the folder 
`/code/AMDMIGraphX` and follow the steps in the above [Build MIGraphX source and install

> libs](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/tree/refine_readme#building-migraphx-source-and-install-libs)

Reviewer 4: Does this link change when it eventually  gets merged in? If not, we need another way to reference the readme

Author: If I remember correctly, it will change accordingly after merging.

Reviewer: I think you can link it as `[Build MIGraphX source and install libs](#building-migraphx-source-and-install-libs)` so it links to itself:

https://stackoverflow.com/a/16426829/375343

Reviewer 4: @scxiao are we changing this based off @pfultz2's suggestion?

Author: Thanks, fixed.

##### Installing the prerequisites

For convenience, the prerequisites can be built automatically with rbuild as:

> ```rbuild build -d depend --cxx=/opt/rocm/llvm/bin/clang++```

Reviewer: There should be a newline after the triple backticks

## Using docker

> The easiest way to setup the development environment is to use docker. With the docker file, you can build an docker image as:

Reviewer: It should be 'a docker image'.

Author: changed.

## Using docker

The easiest way to setup the development environment is to use docker. Docker files for different versions of ROCm are

> provided. ```hcc.docker``` is for ROCm3.3, and ```hip-clang.docker``` is for ROCm3.7.

Reviewer: Users should just use the top-level Dockerfile. hcc is deprecated and these `hcc.docker` and `hip-clang.docker` will be removed in the future.

Author: Containers built from these two files work fine, while a container built from the Dockfile does not work, some components are missed.

Reviewer: Why does the Dockerfile not work? We use the Dockerfile on our CI, so it does work.
